core/name.c

```diff
@@ -0,0 +1,290 @@
+/*
+ * Copyright (c) 2018 Kryptos Logic
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *   1. Redistributions of source code must retain the above copyright notice,
+ *      this list of conditions and the following disclaimer.
+ *
+ *   2. Redistributions in binary form must reproduce the above copyright
+ *      notice, this list of conditions and the following disclaimer in the
+ *      documentation and/or other materials provided with the distribution.
+ *
+ *   3. Neither the name of the copyright holder nor the names of its
+ *      contributors may be used to endorse or promote products derived from
+ *      this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "include/name.h"
+#include "include/vmx.h"
+
+const char *name_vmcs_component(int value)
+{
+    // Intel SDM Vol. 3C: Table 24-17. Structure of VMCS Component Encoding
+#define HASH(x) \
+    (((x) & 0x003E) >> 1) /* Index */ | \
```

Comment:
nit: It's a brilliant idea to omit the redundant Access Type bit as well as the upper 4 bits of Index, but the latter took me some time to confirm... How about adding a brief explanation, something like:

> As of this writing, no VMCS component is encoded with an Index of more than 5 bits (Intel SDM Vol. 3D, Appendix B: Field Encoding in VMCS), even though 9 bits are allocated to Index.

Also nit: Perhaps you could replace the 5 with a macro (e.g. `#define VMCS_ENC_INDEX_BITS 5`). That way, in case we ever need a longer Index, we just need to update a single line of code.

Reply:
Fully agree with the first point.

As for adding the macro: I think it could make the code slightly more unreadable:
```diff
#define VMCS_ENC_INDEX_BITS 5
-(((x) & 0x003E) >> 1) /* Index */ | \
+(((x) & (((1 << VMCS_ENC_INDEX_BITS) - 1) << 1) >> 1
```
Promoting 0x003E (5-bit) to 0x007E (6-bit) or 0x00FE (7-bit) or 0x01FE (8-bit) feels more intuitive to me, since we can directly observe the `component_index_t` values, and simply "see" if our mask is compact enough. In fact, that's how I came up with 0x3E in the first place.

Just thought: If you still stand by the macro, I'll apply the inlined patch above.

Reply:
> [...] since we can directly observe the component_index_t values, and simply "see" if our mask is compact enough.

Ah, that's an interesting perspective. I was going through all the tables in Appendix B and focusing on the Index column... Sure, please keep the hardcoded bitmasks.